REF: add suggested refactoring support for function change signature

Kobzol opened the pull request:
This PR adds support for the suggested refactoring for function signature changes. Originally it was added in https://github.com/intellij-rust/intellij-rust/pull/5913 just for simple binding renames.

I took inspiration from the Kotlin implementation and tried to reduce it to be as small as possible. So for example default values for parameters are not implemented in this PR.

![suggested-refactoring](https://user-images.githubusercontent.com/4539057/115121622-feef7380-9fb3-11eb-9d06-468fb08d0310.gif)

changelog: Add suggested refactoring support for changing function signatures.

dima74 commented:
Great work!

Some comments:

* Swapping function parameters doesn't swap call arguments:
![Peek 2021-04-26 16-56](https://user-images.githubusercontent.com/6505554/116094612-5bd3e380-a6b0-11eb-9d36-ce23d81978fc.gif)

* In Kotlin suggested function rename can be done without this additional dialog:

dima74 commented:
Great work!

Some comments:

* Swapping function parameters doesn't swap call arguments:
![Peek 2021-04-26 16-56](https://user-images.githubusercontent.com/6505554/116094612-5bd3e380-a6b0-11eb-9d36-ce23d81978fc.gif)

* In Kotlin suggested function rename can be done without this additional dialog:
![image](https://user-images.githubusercontent.com/6505554/116097257-a5bdc900-a6b2-11eb-8054-96820b3c2cf9.png)

* When changing `fn func(b: i32) {}` to `fn func(a: i32, b: i32) {}`, function call `func(1)` is changed to `func(1, )`. Looks like it should be changed to `func(, 1)`

* I receive `PsiInvalidElementAccessException` when trying to change `fn func(b: &str) {}` to `fn func(a: i32, b: &str) {}`

Kobzol commented:
> In Kotlin suggested function rename can be done without this additional dialog:

It also shows a dialog, it's just a different one (doesn't show the diff). You also have to click to confirm it. Do you want me to change the dialog?

![rename](https://user-images.githubusercontent.com/4539057/116140319-740b2900-a6d7-11eb-9ebf-3a44ee09753f.gif)

Kobzol commented:
@dima74 It seems that the `dump` method that I'm using in tests is not available in CLion. What's the proper solution to fix it? Copy-paste the implementation from IDEA?

dima74 commented:
> It also shows a dialog, it's just a different one (doesn't show the diff). You also have to click to confirm it. Do you want me to change the dialog?

Oh, It shows dialog when invoked from gutter icon, but doesn't show dialog when invoked from context actions menu:

![Peek 2021-04-28 21-50](https://user-images.githubusercontent.com/6505554/116457178-bfeed700-a86b-11eb-9e77-27104d6c17bf.gif)


> It seems that the dump method that I'm using in tests is not available in CLion. What's the proper solution to fix it? Copy-paste the implementation from IDEA?

Unfortunatelly looks like copying is the only option

---

* I receive `PsiInvalidElementAccessException` when trying to change `fn func() {}` to `fn func<'a>(a: &'a str) {}`. Btw looks like we currently touch/change PSI of function declaration somehow, can we keep function declaration as is (since user already made all necessary changes) and change only function calls?

Kobzol commented:
> Oh, It shows dialog when invoked from gutter icon, but doesn't show dialog when invoked from context actions menu:

Ah, nice catch. I changed it so that if only the name is changed, the simpler name refactoring will be used.

> Btw looks like we currently touch/change PSI of function declaration somehow, can we keep function declaration as is (since user already made all necessary changes) and change only function calls?

Interesting. Based on this comment from `SuggestedRefactoringExecution::performChangeSignature`:
```kotlin
/**
   * Performs Change Signature refactoring. This method is invoked with the declaration reverted to its original state
   * before user changes (the old signature). The list of imports is also reverted to the original state.
   */
  abstract fun performChangeSignature(data: SuggestedChangeSignatureData, newParameterValues: List<NewParameterValue>, preparedData: Any?)
```

I thought that the refactoring reverts the declaration to the original state and we actually have to change it. But it seems that even if I don't run the change signature refactoring, it keeps the changed declaration. I changed it so that the original function is not touched.

dima74 commented:
Also please squash commits into one

Kobzol commented on src/main/kotlin/org/rust/ide/refactoring/suggested/RsSuggestedRefactoringUI.kt:
```diff
@@ -0,0 +1,40 @@
+
+        val newParameters = data.newSignature.parameters
+        val dropLastN = newParameters.reversed().count { isNewParameter(it) }
+        for (parameter in newParameters.dropLast(dropLastN)) {
```
Oops, this was actually leftover code copied from Kotlin. Adding support for default values seemed to be quite easy, so I smuggled it into this PR.

Kobzol commented:
I added support for default values and squashed commits.

When trying the refactoring, I got the following exception, which happened when I tried to change function signature during index update. Should I
1) Avoid using `getValueParameters` altogether in that method, or
2) Do not build the signature when index is updating (I think this is called dumb mode?)

```
com.intellij.openapi.project.IndexNotReadyException: Please change caller according to com.intellij.openapi.project.IndexNotReadyException documentation
	at com.intellij.openapi.project.IndexNotReadyException.create(IndexNotReadyException.java:67)
	at com.intellij.openapi.project.IndexNotReadyException.create(IndexNotReadyException.java:62)
	at org.rust.openapiext.UtilsKt.checkIsSmartMode(utils.kt:103)
	at org.rust.lang.core.resolve2.FacadeUpdateDefMapKt.getOrUpdateIfNeeded(FacadeUpdateDefMap.kt:39)
	at org.rust.lang.core.resolve2.FacadeResolveKt.findModDataFor(FacadeResolve.kt:629)
	at org.rust.lang.core.psi.RsFile.doGetCachedData(RsFile.kt:103)
	at org.rust.lang.core.psi.RsFile.access$doGetCachedData(RsFile.kt:68)
	at org.rust.lang.core.psi.RsFile$cachedData$1$value$1.compute(RsFile.kt:93)
	at org.rust.lang.core.psi.RsFile$cachedData$1$value$1.compute(RsFile.kt:68)
	at com.intellij.openapi.util.RecursionManager$1.computePreventingRecursion(RecursionManager.java:111)
	at com.intellij.openapi.util.RecursionGuard.doPreventingRecursion(RecursionGuard.java:43)
	at com.intellij.openapi.util.RecursionManager.doPreventingRecursion(RecursionManager.java:68)
	at org.rust.openapiext.UtilsKt.recursionGuard(utils.kt:67)
	at org.rust.openapiext.UtilsKt.recursionGuard$default(utils.kt:66)
	at org.rust.lang.core.psi.RsFile$cachedData$1.compute(RsFile.kt:93)
	at com.intellij.psi.util.CachedValuesManager$1.compute(CachedValuesManager.java:158)
	at com.intellij.psi.impl.PsiCachedValueImpl.doCompute(PsiCachedValueImpl.java:54)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$3(CachedValueBase.java:227)
	at com.intellij.util.CachedValueBase.computeData(CachedValueBase.java:42)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$4(CachedValueBase.java:227)
	at com.intellij.openapi.util.RecursionManager$1.computePreventingRecursion(RecursionManager.java:111)
	at com.intellij.openapi.util.RecursionGuard.doPreventingRecursion(RecursionGuard.java:43)
	at com.intellij.openapi.util.RecursionManager.doPreventingRecursion(RecursionManager.java:68)
	at com.intellij.util.CachedValueBase.getValueWithLock(CachedValueBase.java:228)
	at com.intellij.psi.impl.PsiCachedValueImpl.getValue(PsiCachedValueImpl.java:43)
	at com.intellij.util.CachedValuesManagerImpl.getCachedValue(CachedValuesManagerImpl.java:72)
	at com.intellij.psi.util.CachedValuesManager.getCachedValue(CachedValuesManager.java:155)
	at org.rust.lang.core.psi.RsFile.getCachedData(RsFile.kt:92)
	at org.rust.lang.core.psi.RsFile.getCrate(RsFile.kt:76)
	at org.rust.lang.core.psi.ext.RsElementKt.getContainingCrate(RsElement.kt:61)
	at org.rust.lang.core.stubs.common.RsAttributeOwnerPsiOrStub.getContainingCrate(PsiOrStubInterfaces.kt:45)
	at org.rust.lang.core.psi.ext.RsDocAndAttributeOwnerKt.evaluateCfg(RsDocAndAttributeOwner.kt:435)
	at org.rust.lang.core.psi.ext.RsDocAndAttributeOwnerKt.evaluateCfg$default(RsDocAndAttributeOwner.kt:424)
	at org.rust.lang.core.psi.ext.RsDocAndAttributeOwnerKt.isEnabledByCfgSelf(RsDocAndAttributeOwner.kt:398)
	at org.rust.lang.core.psi.ext.RsDocAndAttributeOwnerKt.getExistsAfterExpansionSelf(RsDocAndAttributeOwner.kt:406)
	at org.rust.lang.core.psi.ext.RsFunctionKt.getValueParameters(RsFunction.kt:73)
	at org.rust.ide.refactoring.suggested.RsSuggestedRefactoringStateChanges.signature(RsSuggestedRefactoringStateChanges.kt:47)
	at com.intellij.refactoring.suggested.SuggestedRefactoringStateChanges.createInitialState(SuggestedRefactoringStateChanges.kt:45)
	at com.intellij.refactoring.suggested.SuggestedRefactoringChangeCollector.editingStarted(SuggestedRefactoringChangeCollector.kt:27)
	at com.intellij.refactoring.suggested.SuggestedRefactoringChangeListener.processBeforeFirstChangeWithPsiAndDocumentInSync(SuggestedRefactoringChangeListener.kt:142)
	at com.intellij.refactoring.suggested.SuggestedRefactoringChangeListener.access$processBeforeFirstChangeWithPsiAndDocumentInSync(SuggestedRefactoringChangeListener.kt:22)
	at com.intellij.refactoring.suggested.SuggestedRefactoringChangeListener$MyDocumentListener.beforeDocumentChange(SuggestedRefactoringChangeListener.kt:210)
	at jdk.internal.reflect.GeneratedMethodAccessor121.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at com.intellij.util.EventDispatcher.dispatchVoidMethod(EventDispatcher.java:123)
	at com.intellij.util.EventDispatcher.lambda$createMulticaster$1(EventDispatcher.java:86)
	at com.sun.proxy.$Proxy77.beforeDocumentChange(Unknown Source)
	at com.intellij.openapi.editor.impl.DocumentImpl.beforeChangedUpdate(DocumentImpl.java:877)
	at com.intellij.openapi.editor.impl.DocumentImpl.updateText(DocumentImpl.java:815)
	at com.intellij.openapi.editor.impl.DocumentImpl.deleteString(DocumentImpl.java:576)
	at com.intellij.openapi.editor.EditorModificationUtil.deleteSelectedText(EditorModificationUtil.java:50)
	at com.intellij.openapi.editor.actions.DeleteSelectionHandler.lambda$executeWriteAction$0(DeleteSelectionHandler.java:41)
	at com.intellij.openapi.editor.impl.CaretModelImpl.lambda$runForEachCaret$3(CaretModelImpl.java:312)
	at com.intellij.openapi.editor.impl.CaretModelImpl.doWithCaretMerging(CaretModelImpl.java:421)
	at com.intellij.openapi.editor.impl.CaretModelImpl.runForEachCaret(CaretModelImpl.java:321)
	at com.intellij.openapi.editor.impl.CaretModelImpl.runForEachCaret(CaretModelImpl.java:296)
	at com.intellij.openapi.editor.actions.DeleteSelectionHandler.executeWriteAction(DeleteSelectionHandler.java:42)
	at com.intellij.openapi.editor.actionSystem.EditorWriteActionHandler$1.run(EditorWriteActionHandler.java:52)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:959)
	at com.intellij.openapi.editor.actionSystem.EditorWriteActionHandler.doExecute(EditorWriteActionHandler.java:66)
	at com.intellij.openapi.editor.actionSystem.DynamicEditorActionHandler.doExecute(DynamicEditorActionHandler.java:52)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.lambda$execute$4(EditorActionHandler.java:199)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.doIfEnabled(EditorActionHandler.java:89)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.execute(EditorActionHandler.java:198)
	at com.intellij.openapi.editor.actionSystem.EditorAction.lambda$actionPerformed$0(EditorAction.java:89)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:216)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:172)
	at com.intellij.openapi.editor.actionSystem.EditorAction.actionPerformed(EditorAction.java:98)
	at com.intellij.openapi.editor.actionSystem.EditorAction.actionPerformed(EditorAction.java:73)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.lambda$performActionDumbAware$5(ActionUtil.java:273)
	at com.intellij.util.SlowOperations.lambda$allowSlowOperations$0(SlowOperations.java:77)
	at com.intellij.util.SlowOperations.allowSlowOperations(SlowOperations.java:68)
	at com.intellij.util.SlowOperations.allowSlowOperations(SlowOperations.java:76)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.performActionDumbAware(ActionUtil.java:273)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher$1.performAction(IdeKeyEventDispatcher.java:616)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.lambda$processAction$3(IdeKeyEventDispatcher.java:676)
	at com.intellij.openapi.application.TransactionGuardImpl.performUserActivity(TransactionGuardImpl.java:95)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.processAction(IdeKeyEventDispatcher.java:675)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.processAction(IdeKeyEventDispatcher.java:626)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.processActionOrWaitSecondStroke(IdeKeyEventDispatcher.java:486)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.inInitState(IdeKeyEventDispatcher.java:481)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.dispatchKeyEvent(IdeKeyEventDispatcher.java:232)
	at com.intellij.ide.IdeEventQueue.dispatchKeyEvent(IdeEventQueue.java:889)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:833)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$8(IdeEventQueue.java:449)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:808)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$9(IdeEventQueue.java:448)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:781)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:502)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```

dima74 commented:
> When trying the refactoring, I got the following exception, which happened when I tried to change function signature during index update

I think we can use `rawValueParameters` instead of `valueParameters`, since changing signature of function which has cfg-disabled parameters is unsupported

Kobzol commented:
Oh, you're completely right! And that also led me to disabling the suggested refactoring if the function has cfg-disabled parameters. I included it in a separate commit for review, since it's not a minor change. I'll squash it if you find it OK.

dima74 commented:
Great! Cfg-disabled parameters check is OK.

Also I found a bug with default parameters - when using some constant as default parameter, import is also added to mod with function declaration, but it should be added only to modules with function invocations. I would suggest extracting default values into separate PR for easier review process (since this PR is already quite big)

Kobzol commented:
Ok, I squashed commits and took out default values. I'll send a new PR with default values, I'll also try to add better support for testing them.

dima74 commented:
Thanks! 👍
bors r+

Kobzol commented:
@dima74 I couldn't simulate the situation that you have described with the import of default value. Could you please post the corresponding code snippet?

dima74 commented:
Sure, something like this, and I tried to add parameter with default value `FOO` (I accept completion variant in default value code fragment)
```rust
fn func() {}

mod mod0 {
    pub const FOO: i32 = 0;
}
mod mod1 {
    fn main() {
        crate::func();
    }
}
```

Kobzol commented:
I see, thank you.
